traffic_ops/testing/api/v4/deliveryservices_test.go

```diff
@@ -1238,6 +1254,272 @@ func VerifyPaginationSupportDS(t *testing.T) {
+			}
+			firstDS.TenantID = &tenant.ID
+			qparams := url.Values{}
+			qparams.Set("tenant", strconv.Itoa(*firstDS.TenantID))
+			resp, _, err := TOSession.GetDeliveryServicesV4(nil, qparams)
+			if err != nil {
+				t.Fatalf("Error in Getting Deliveryservice by Tenant:%v - %v", err, resp)
+			}
+			if len(resp) == 0 {
+				t.Errorf("No delivery service available for the Tenant %v", *firstDS.CDNName)
+			} else {
+				if *resp[0].Tenant != *firstDS.Tenant {
+					t.Fatalf("name expected: %s, actual: %s", *firstDS.Tenant, *resp[0].Tenant)
+				}
+			}
+		}
+	}
+}
+
+func GetDeliveryServiceByInvalidTenant(t *testing.T) {
+	qparams := url.Values{}
+	qparams.Set("tenant", "10000")
+	resp, _, err := TOSession.GetDeliveryServicesV4(nil, qparams)
+	if err != nil {
+		t.Fatalf("Error!! Getting Deliveryservice by Invalid Tenant ID %v", err)
+	}
+	if len(resp) >= 1 {
+		t.Fatalf("Error!! Invalid Tenant shouldn't have any response %v Error %v", resp, err)
+	}
+}
+
+func GetDeliveryServiceByValidType(t *testing.T) {
+	if len(testData.DeliveryServices) > 0 {
+		firstDS := testData.DeliveryServices[0]
+
+		if firstDS.TypeID == nil && firstDS.Type != nil {
+			ty, _, err := TOSession.GetTypeByNameWithHdr(firstDS.Type.String(), nil)
+			if err != nil {
+				t.Fatalf("Error in Getting Type by Name: %v", err)
+			}
+			if len(ty) == 0 {
+				t.Fatalf("no Type named %v" + firstDS.Type.String())
+			}
+			firstDS.TypeID = &ty[0].ID
+			qparams := url.Values{}
+			qparams.Set("type", strconv.Itoa(*firstDS.TypeID))
+			resp, _, err := TOSession.GetDeliveryServicesV4(nil, qparams)
+			if err != nil {
+				t.Fatalf("Error in Getting Deliveryservice by Type:%v - %v", err, resp)
+			}
+			if len(resp) == 0 {
+				t.Errorf("No delivery service available for the Type %v", *firstDS.CDNName)
+			} else {
+				if *resp[0].Type != *firstDS.Type {
+					t.Fatalf("Type expected: %s, actual: %s", *firstDS.Type, *resp[0].Type)
+				}
+			}
+		}
+	}
+}
+
+func GetDeliveryServiceByInvalidType(t *testing.T) {
+	qparams := url.Values{}
+	qparams.Set("type", "10000")
+	resp, _, err := TOSession.GetDeliveryServicesV4(nil, qparams)
+	if err != nil {
+		t.Fatalf("Error!! Getting Deliveryservice by Invalid Type ID %v", err)
+	}
+	if len(resp) >= 1 {
+		t.Fatalf("Error!! Invalid Type shouldn't have any response %v Error %v", resp, err)
+	}
+}
+
+func GetDeliveryServiceByInvalidAccessibleTo(t *testing.T) {
+	qparams := url.Values{}
+	qparams.Set("accessibleTo", "10000")
+	resp, _, err := TOSession.GetDeliveryServicesV4(nil, qparams)
+	if err != nil {
+		t.Fatalf("Error!! Getting Deliveryservice by Invalid AccessibleTo %v", err)
+	}
+	if len(resp) >= 1 {
+		t.Fatalf("Error!! Invalid AccessibleTo shouldn't have any response %v Error %v", resp, err)
+	}
+}
+
+func GetDeliveryServiceByValidXmlId(t *testing.T) {
+	if len(testData.DeliveryServices) > 0 {
+		firstDS := testData.DeliveryServices[0]
+
+		resp, _, err := TOSession.GetDeliveryServiceByXMLIDNullableWithHdr(*firstDS.XMLID, nil)
```

Comment:
if `firstDS.XMLID` is `nil`, this will segfault

Reply:
Handled Nil values